plugin/utils/shell.py

```diff
@@ -113,145 +124,194 @@ def generate_rig_edit():
-            logging.info("Angle_radians: "+str(angle_radians))
-            logging.info("Angle_degrees: "+str(angle_degrees))
-            
-            #Calculate the rotation matrix
-            angle = angle_radians
-            #Calculate the axis
-            axis = vector1.cross(vector2)
-            #Log axis
-            logging.info("Axis: "+str(axis))
-            #Create the rotation matrix
-            rotation_matrix = Matrix.Rotation(-angle_radians, 4, axis)
-            # Apply the final matrix to the pose bone
-            p_bone.parent.matrix = Matrix.Translation(+origin) @ rotation_matrix @ Matrix.Translation(+origin).inverted() @ p_bone.parent.matrix
-            
-            #------------------------------------
-            
+        #We check if any bones have scale and warn the user, as clearing their scale will significantly alter the result.
+        if not vectorisclose(p_bone.scale,Vector((1,1,1)),0.0001):
+            logging.info(f"Bone has scale. Value = {repr(p_bone.scale)}, difference: {(p_bone.scale - Vector((1,1,1))).length}")
+            msgs.append(f"Warning: {str(p_bone.name)} had scale transforms. Reset scale for all bones and try again.")
+            return msgs
+        
+        #We check for NODE bones with transforms and skip them.
+        if (not vectorisclose(p_bone.location,Vector((0,0,0)),0.0001) or not vectorisclose(p_bone.scale,Vector((1,1,1)),0.0001) or not vectorisclose(Vector(p_bone.rotation_quaternion),Vector((1,0,0,0)),0.0001)) and p_bone.name.startswith("NODE_"):
+            #Ignore posed NODE bones and proceed to the next, their offsets can be applied directly.
             editnumber = editnumber + 1
+            logging.info(f"rig edit number {editnumber}")
+            logging.info(f"NODE with offsets detected. Applying offsets directly.")
+            continue
+        
+        #We append any remaining bones that have been posed.
+        if (not vectorisclose(p_bone.location,Vector((0,0,0)),0.0001) or not vectorisclose(p_bone.scale,Vector((1,1,1)),0.0001) or not vectorisclose(Vector(p_bone.rotation_quaternion),Vector((1,0,0,0)),0.0001)):
+            #Append the bones to the list of posed bones.
+            posedbones_list.append(p_bone)
+    
+    
+    #Iterate over every posed bone.
+    for p_bone in posedbones_list:
+        #Update bone locations
+        bpy.context.view_layer.update()
+            
+        #Debug info
+        logging.info(f"rig edit number {editnumber+1}")
+        logging.info(f"name: {p_bone.name}")
+        logging.info(f"parent: {p_bone.parent.name}")
+        
+        #Get pose mode position of bonebase
+        bonebase_pos_pose = Vector(p_bone.matrix.translation)
+        
+        #Reset position
+        p_bone.location = Vector((0, 0, 0))
+        #Update location after moving.
+        bpy.context.view_layer.update()
+        
+        #Get rest mode position of bonebase
+        bonebase_pos_rest = Vector(p_bone.matrix.translation)
```

Comment:
typo - same as bonebase_pos_pose ?

Reply:
It's stored before and after clearing the pose. So I have the global location of the rest position and the pose position, so i can calculate the delta offset. This way i don't run into a problem where it adds up the offsets depending on the order that you calculate the offsets like with the first version of the script that caused bones to get incorrectly positioned after several chains of parents

Reply:
ah ok I see